### PR TITLE
Support keyed children

### PR DESCRIPTION
**Pull request by fitzgen**
@alexcrichton @tschneidereit: I'd appreciate some eyes on this PR (particularly the third commit) :eyes: 

Using keyed children gives users more control over whether particular physical nodes are reused, moved, deleted, or created afresh. Specifically, if the old virtual node and the new virtual node have the same key, then the physical DOM node must be reused, and if they don't then the old physical DOM node cannot be reused.

Presumably there is some sort of CSS transition animation or third party JS that makes the virtual DOM diffing algorithm observable, or else users would let the runtime do whatever it thought was best.

This is loosely based on Inferno's keyed patching implementation. However, we have to modify the algorithm since we are compiling the diff down into change list instructions that will be executed later, rather than applying the changes to the DOM directly as we compare virtual DOMs.

https://github.com/infernojs/inferno/blob/36fd96/packages/inferno/src/DOM/patching.ts#L530-L739

### CODE REVIEW

File: src/builder.rs [hunk 1]
```diff
@@ -220,6 +226,56 @@ where
...
+        } else {
+            self.flags &= !ElementNodeFlags::HAS_KEYED_CHILDREN;
+        }
+        self
+    }
+
+    /// Set this element's key. The key must be less than or equal to `2^31 -
+    /// 1`.
+    ///
+    /// This element's parent must have `has_keyed_children(true)` set.
```
Comment (alexcrichton) [inline src/builder.rs]: I'm still likely naive at this point (I don't currently have a grasp on what keyed children are used for) but it seems unfortunate from an API-design perspective that this property needs to be upheld. 

Would it be possible to have debug asserts for this though throughout just to make sure test builds try to catch this?

File: src/builder.rs [hunk 2]
```diff
@@ -220,6 +226,56 @@ where
...
+        if has_keyed_children {
+            self.flags |= ElementNodeFlags::HAS_KEYED_CHILDREN;
+        } else {
+            self.flags &= !ElementNodeFlags::HAS_KEYED_CHILDREN;
+        }
+        self
+    }
+
+    /// Set this element's key. The key must be less than or equal to `2^31 -
+    /// 1`.
```
Comment (alexcrichton) [inline src/builder.rs]: Perhaps mention that in layman's terms this means the top bit can't be used?

File: src/diff.rs [hunk 3]
```diff
@@ -2,40 +2,47 @@ use crate::{
...
 
+// Diff the `old` node with the `new` node. Emits instructions to modify a
+// physical DOM node that reflects `old` into something that reflects `new`.
+//
+// Upon entry to this function, the physical DOM node must be on the top of the
+// stack:
+//
+//     [... node]
+//
+// The stack is in the same state when this function exits.
```
Comment (alexcrichton) [inline src/diff.rs]: Without having a deep understanding of dodrio, I'm not entirely certain what stack this is referring to. Is this the stack in the change list builder which is used as sort of a stack machine for all the opcodes generated? (I presume so, just wanted to confirm!)

File: src/diff.rs [hunk 4]
```diff
@@ -2,40 +2,47 @@ use crate::{
...
+    old: &Node,
+    new: &Node,
     cached_roots: &mut FxHashSet<CacheId>,
 ) {
     match (&new.kind, &old.kind) {
         (
             &NodeKind::Text(TextNode { text: new_text }),
             &NodeKind::Text(TextNode { text: old_text }),
         ) => {
-            debug!("  both are text nodes");
```
Comment (alexcrichton) [inline src/diff.rs:23]: Did these end up generating too much traffic in the debug logs? (just curious as to the removal)

File: src/diff.rs [hunk 5]
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
...
                 change_list,
                 registry,
-                new_child.clone(),
+                &new[old.len()..],
                 cached_roots,
             );
-            change_list.append_child();
+        }
+        Ordering::Equal => {
+            change_list.pop();
```
Comment (alexcrichton) [inline src/diff.rs:795]: I was trying to pair this up with another "push" in this function, and it took me a bit following it to figure out what was going on. Could the `i == 0` case be pulled out of the loop and listed inline above? I think the `i == 0` is guaranteed to fun due to the nonempty assertions for both arrays, but it may be a bit more clearer to pull it out to see what push this is matching.

I also don't really know what's being pushed and popped, so feel free to say "not necessary" :)

File: src/diff.rs [hunk 6]
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
...
+// When this function returns, the stack is in the same state.
+fn remove_all_children(
+    change_list: &mut ChangeListBuilder,
+    registry: &mut EventsRegistry,
+    old: &[Node],
+) {
+    for child in old {
+        registry.remove_subtree(child);
+    }
+    // Fast way to remove all children: set the node's textContent to an empty
```
Comment (alexcrichton) [inline src/diff.rs:836]: Learn something new every day!

File: src/diff.rs [hunk 7]
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
...
+    // value, so if we have that many new nodes, it won't work. This
+    // check is a bit academic (hence only enabled in debug), since
+    // wasm32 doesn't have enough address space to hold that many nodes
+    // in memory.
+    debug_assert!(new.len() < u32::MAX as usize);
+
+    // Map from each `old` node's key to its index within `old`.
+    let mut old_key_to_old_index = FxHashMap::default();
+    old_key_to_old_index.reserve(old.len());
+    old_key_to_old_index.extend(old.iter().enumerate().map(|(i, o)| (o.key(), i)));
```
Comment (alexcrichton) [inline src/diff.rs:488]: This brings up an interesting idea in the sense that it's the onus of the user to not reuse keys, right?

(should that be added to the `key` docs?

Comment (alexcrichton) [reply src/diff.rs:488]: Actually now I'm slightly confused, it looks like `u32::MAX` is used for non-present keys, so that just means that `u32::MAX` as a key in this map is overwritten multiple times? I guess though non-max values are supposed to be unique?

File: src/diff.rs [hunk 8]
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
...
+            i
+        } else {
+            u32::MAX as usize
+        }
+    }));
+
+    // If none of the old keys are reused by the new children, then we
+    // remove all the remaining old children and create the new children
+    // afresh.
+    if shared_suffix_count == 0 && shared_keys.is_empty() {
```
Comment (alexcrichton) [inline src/diff.rs:508]: This condition is throwing me a bit for a spin, I am naively expecting it to *just* be checking the `shared_keys` set, but I can't quite figure out why `shared_suffix_count` is factored in here?

Should `shared_suffix_count == 0` only be used to optimize which opcodes are emitted? I'd expect that we'd still have some optimized path for shared `shared_suffix_count` is nonzero but there's no shared keys

File: src/diff.rs [hunk 9]
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
...
+        } else {
+            start = end + 1;
+        }
+    }
+
+    // Remove any old children whose keys were not reused in the new
+    // children. Remove from the end first so that we don't mess up indices.
+    let mut removed_count = 0;
+    for (i, old_child) in old.iter().enumerate().rev() {
+        if !shared_keys.contains(&old_child.key()) {
```
Comment (alexcrichton) [inline src/diff.rs:554]: Hm ok I may again be very confused, but this sort of looks like it's preserving everything with a key of `u32::MAX`?

**Review by alexcrichton — COMMENTED** [review]
Ok so I followed everything pretty well up to the point that I hit `diff_keyed_middle`. Man that function is gnarly! I'm gonna leave the comments I have for now, but I suspect you'll have some comments which help clear up a few things for me and I can take another stab later to finish it :)

File: src/builder.rs (hunk 1 above)
Comment (fitzgen) [reply src/builder.rs]: Yeah I can def add debug asserts. It is a somewhat unfortunate constraint, but my hope is that `typed-html` or whatever would do this automatically 99% of the time.

The reason it is this way is because checking for keys is an O(n) scan, but the person rendering the virtual dom tree just *knows* whether they are using keys or not, so it seems absolutely silly to do that linear search.

But also if there are keyed children, then they should all be keyed, so probably we should do:

* fast check for if first child is keyed
* debug assert that all children are keyed or all children are not keyed
* remove the has-keyed-children flag, `bitflags` dep, and `ElementNodeFlags`
* just store a `u32` key in `ElementNode` and remove the you-can't-use-the-top-bit restriction

File: src/diff.rs (hunk 3 above)
Comment (fitzgen) [reply src/diff.rs]: Yeah this is the interpreter's stack if it has evaluated all the emitted instructions thus far. Happy for suggestions of more precise but less wordy ways of saying this :)

File: src/diff.rs (hunk 4 above)
Comment (fitzgen) [reply src/diff.rs:23]: Yeah. Essentially, I found that the only logs I really care about are which instructions get emitted.

File: src/diff.rs (hunk 8 above)
Comment (fitzgen) [reply src/diff.rs:508]: Your intuition that there should be a fast path regardless of `shared_suffix_count`'s value is correct. The reason it doesn't exist is simply that we don't currently have instructions/functions for mass deleting and inserting in the middle of the children, only all of the children, or the latter half of the children.

It is somethign that I feel can be a follow up PR.

File: src/diff.rs (hunk 5 above)
Comment (fitzgen) [reply src/diff.rs:795]: Yeah because `new` and `old` are non-empty, we know we execute the loop at least once, and therefore push a child onto the stack, and that is what is being popped here. I can make a comment to this effect to clear things up.

File: src/diff.rs (hunk 7 above)
Comment (fitzgen) [reply src/diff.rs:488]: Yeah I should add an assertion for this.

File: src/diff.rs (hunk 8 above)
Comment (fitzgen) [reply src/diff.rs]: Your intuition that there should be a fast path regardless of shared_suffix_count's value is correct. The reason it doesn't exist is simply that we don't currently have instructions/functions for mass deleting and inserting in the middle of the children, only all of the children, or the latter half of the children.

It is somethign that I feel can be a follow up PR.

File: src/diff.rs (hunk 3 above)
Comment (fitzgen) [reply src/diff.rs]: Calling this "the change list stack" instead of "the stack" to clarify here.

File: src/builder.rs (hunk 1 above)
Comment (fitzgen) [reply src/builder.rs]: Done!

File: src/builder.rs (hunk 2 above)
Comment (fitzgen) [reply src/builder.rs]: Only `u32::MAX` is reserved now, and this is documented and asserted.

**Comment by axelf4** [comment]
Just alerting you to the fact that there exists some overlap here between this and [a diff crate](https://github.com/axelf4/lis/blob/c695f0e5b59ead9d43e0b67a172945e4d07da4f2/src/lib.rs#L244:L336) I wrote.

File: src/diff.rs (hunk 3 above)
Comment (alexcrichton) [reply src/diff.rs]: Nah I think this is fine as it was (I'll probably see an update soon anyway), mostly just wanted to double-check my own understanding!

**Comment by alexcrichton** [comment]
So I started off today trying to get a better idea of the middle diffing function, and ended up spending basically all day following @fitzgen down profiling rabbit holes.

At the beginning of the day though we concluded that the iteration here in the middle diffing can probably be simplified to not use chunks but rather just in-order reverse iteration. I think that's probably the next step here!
